pkg/detector/preemption.go

```diff
@@ -0,0 +1,181 @@
+package detector
+
+import (
+	"context"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/klog/v2"
+
+	policyv1alpha1 "github.com/karmada-io/karmada/pkg/apis/policy/v1alpha1"
+	"github.com/karmada-io/karmada/pkg/features"
+	"github.com/karmada-io/karmada/pkg/util"
+	"github.com/karmada-io/karmada/pkg/util/helper"
+	"github.com/karmada-io/karmada/pkg/util/names"
+)
+
+// preemptionEnabled checks if preemption is enabled.
+func preemptionEnabled(preemption policyv1alpha1.PreemptionBehavior) bool {
+	if preemption != policyv1alpha1.PreemptAlways {
+		return false
+	}
+	if !features.FeatureGate.Enabled(features.PolicyPreemption) {
+		klog.Warningf("Cannot handle the preemption process because feature gate %q is not enabled.", features.PolicyPreemption)
+		return false
+	}
+	return true
+}
+
+// handleClusterPropagationPolicyPreemption handles the preemption process of PropagationPolicy.
+// If preemption is not enabled, the process will be aborted.
+// The preemption rule: high-priority PP > low-priority PP > CPP.
+func (d *ResourceDetector) handlePropagationPolicyPreemption(policy *policyv1alpha1.PropagationPolicy) error {
+	if !preemptionEnabled(policy.Spec.Preemption) {
+		return nil
+	}
+
+	var errs []error
+	for _, rs := range policy.Spec.ResourceSelectors {
+		resourceTemplate, err := d.fetchResourceTemplate(rs)
+		if err != nil || resourceTemplate == nil {
```

Comment:
if resource template == nil, it seems that will append a nil to the slice.

Reply:
it's ok even if err is `nil`.
```
// NewAggregate converts a slice of errors into an Aggregate interface, which
// is itself an implementation of the error interface.  If the slice is empty,
// this returns nil.
// It will check if any of the element of input error list is nil, to avoid
// nil pointer panic when call Error().
func NewAggregate(errlist []error) Aggregate {
	if len(errlist) == 0 {
		return nil
	}
	// In case of input error list contains nil
	var errs []error
	for _, e := range errlist {
		if e != nil {
			errs = append(errs, e)
		}
	}
	if len(errs) == 0 {
		return nil
	}
	return aggregate(errs)
}
```

Reply:
Got it, if resource template == nil, it seems that we do not need to retry.

Reply:
sure, if `resourceTemplate` is not found or being deleted, `err` is `nil`.

Reply:
I know `NewAggregate` can ignore `nil` errors, but from a code readability perspective, we can prevent nil from the errs slice.